### PR TITLE
code generation from the ast

### PR DESCRIPTION
**Pull request by eyberg**
This is about auto-generating code from the AST to wrap goroutines and fixing blank identifer errors.

### CODE REVIEW

**Comment by nathany** [comment]
I can take a look and leave some comments, sure.

File: content/code-generation-from-the-ast.md [hunk 1]
```diff
@@ -0,0 +1,287 @@
++++
+author = ["Ian Eyberg"]
+date = "2015-03-13T12:50:47-07:00"
+linktitle = "Code Generation From the AST"
+title = "Code Generation From the AST"
++++
+
+[Deprehend](https://github.com/deferpanic/deprehend "Deprehend") is a tool
+that auto-generates go src to do things like wrapping goroutines with
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: Suggestion: ~~to do things like wrapping~~ to wrap

The precedence isn't entirely clear to me. Is it? 

1. (goroutines with panic handlers) and (http handlers for metrics tracking)
2. (goroutines with panic handlers and http handlers) for metrics tracking

File: content/code-generation-from-the-ast.md [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+panic handlers and http handlers for metrics tracking.
+
+![](/postimages/code-generation-from-the-ast/deprehend.png)
+
+Over at [DeferPanic](https://deferpanic.com "DeferPanic") we do a lot of
+application monitoring specifically for you, go engineers.
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: Go engineers.

(capitalize)

File: content/code-generation-from-the-ast.md [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+are an anti-magic crowd. To this end, we have exposed most of our
+functionality as wrappers around the stdlib. What this means is that you
+have to manually add code to use the client library effectively.
+
+Engineers are a lazy bunch - it's why we are engineers. We like to
+automate as much as possible.
+
+As more and more companies started integrating with us we kept hearing
+more and more of - “You mean I have to wrap every single http handler
+and every single go routine?”
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: goroutine

(one word)

File: content/code-generation-from-the-ast.md [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+
+This reminded gophers of their initial reaction to error handling in go
+- "You mean I have to wrap every function with error handling?” :)
+
+For smaller codebases or new projects this isn't a big deal but for
+existing codebases this can be a non-trivial amount of work and what
+happens if you forget something?
+
+We admit, this is not only a complete pain in the ass but it is hard to
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: think of the children :-P

File: content/code-generation-from-the-ast.md [hunk 5]
```diff
@@ -0,0 +1,287 @@
...
+scale as you have no way of knowing new code has been handled correctly
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: knowing <ins>if</ins> new code

File: content/code-generation-from-the-ast.md [hunk 6]
```diff
@@ -0,0 +1,287 @@
...
+or not - it kinda defeats the purpose of handling un-expected
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: unexpected needs no hyphen

File: content/code-generation-from-the-ast.md [hunk 7]
```diff
@@ -0,0 +1,287 @@
...
+errors/panics.
+
+So we started hacking out Deprehend .
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: remove extra space before period

File: content/code-generation-from-the-ast.md [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+
+What do we mean by the correctness? Have you ever tried to build your
+program via go build and it fails to compile? Maybe like me you use
+[Fresh](https://github.com/pilu/fresh "Fresh") and you have a OCD of
+hitting save every 2-3 seconds as you are thinking - you'll probably get
+a few failed builds here and there as you save if you have only written
+half an expression. This is the compiler failing to produce a correct
+AST of which it complains and then bails out.
+
+What about the symbol table? The symbol table is another data structure
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: Interesting. I thought Go didn't even have a symbol table?

> "the language has been designed to be easy to analyze and can be parsed without a symbol table." [Go FAQ](http://golang.org/doc/faq#different_syntax)

File: content/code-generation-from-the-ast.md [hunk 9]
```diff
@@ -0,0 +1,287 @@
...
+used by the compiler where every variable you declare is complemented
+with meta-data such as it's type (int vs string). In some languages you
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: ~~it's~~ its type

(no apostrophe unless it's a contraction for "it is")

File: content/code-generation-from-the-ast.md [hunk 10]
```diff
@@ -0,0 +1,287 @@
...
+
+Then the parser might conclude that this forms an assignment expression
+where i evaluates to 5 with a resulting tree looking similar to:
+
+![](/postimages/code-generation-from-the-ast/ast.png)
+
+To be clear this is just a general example - not necessarily what you'll
+find in any given language.
+
+## The Purpose of it All:
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: sentence case vs. title case is inconsistent for headers.

File: content/code-generation-from-the-ast.md [hunk 11]
```diff
@@ -0,0 +1,287 @@
...
+
+With the structure of an AST we can now determine what exactly the
+expression is versus what we (humans) read it to be. This is a very important
+distinction to make.
+
+One of the main functions of deprehend is to find and auto-wrap errors
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: Deprehend is capitalized elsewhere.

File: content/code-generation-from-the-ast.md [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+
+### Problem 2:
+
+What if err is not an error at all? What if it's a string? We simply
+don't know by parsing the source with a regex. There is no affirmative
+way to understand that without context.
+
+### Problem 3:
+
+Since our code replacement is potentially adding more lines how do we do
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: Suggest adding a comma: more lines, how do we...

File: content/code-generation-from-the-ast.md [hunk 13]
```diff
@@ -0,0 +1,287 @@
...
+this without accidentally modifying things we should not be modifying?
+
+The answer:
+
+## We dump the AST.
+
+From the top we'll recursively walk the file tree specified on the
+command line via http://golang.org/pkg/path/filepath/#Walk .
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: Suggestion: make these links read a bit nicer.

[filepath.Walk](http://golang.org/pkg/path/filepath/#Walk)

Comment (nathany) [inline content/code-generation-from-the-ast.md]: Also some extra spaces before periods.

File: content/code-generation-from-the-ast.md [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+
+From there we parse the source code of each file to grab the AST via
+http://golang.org/pkg/go/parser/#ParseFile .
+
+We also grab the type information for our expressions and stuff it
+into a map via  https://godoc.org/golang.org/x/tools/go/types#Info .
+
+We walk the AST in depth first order via
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: Suggestion: hyphenate depth-first

File: content/code-generation-from-the-ast.md [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+http://golang.org/pkg/go/ast/#Inspect . (remember, we are dealing
+w/trees)
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: Suggestion: spell out with

File: content/code-generation-from-the-ast.md [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+
+For each node we look for a few things:
+
+  * does this node contain a go routine?
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: goroutine

(one word)

File: content/code-generation-from-the-ast.md [hunk 17]
```diff
@@ -0,0 +1,287 @@
...
+
+  * does this node make an assignment? if so is the type an error?
+
+
+## Errors
+
+  If an error is found we do a couple of things.
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: I'm not sure why these paragraphs are indented?

File: content/code-generation-from-the-ast.md [hunk 18]
```diff
@@ -0,0 +1,287 @@
...
+not working:
+
+  <u>strconv.ParseInt: parsing "$10": invalid syntax</u>
+
+  Guess what the value of i is now? That's right - [the zero value](https://golang.org/ref/spec#The_zero_value "the zero value") of an int - 0.
+
+  By discovering the places where we purposely or inadvertently throw
+away errors we can uncover potential bugs lurking on our production
+systems and prevent a rash of angry customers wondering why the service
+was free for so long when we should have been charging them $10.
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: :smile:

File: content/code-generation-from-the-ast.md [hunk 19]
```diff
@@ -0,0 +1,287 @@
...
+daemon because someone missed a divide by zero.
+
+By consulting the AST we can confirm what our code is doing rather than just by
+guessing via some regexen.
+
+There's a famous [stackoverflow post](http://stackoverflow.com/questions/1732348/regex-match-open-tags-except-xhtml-self-contained-tags "stackoverflow post")
+where someone asks about using regexen to parse html. Similarly, if you are trying to understand source
+- don't guess - just look it up in the AST.
+
+For an example of how we re-write the source grab deprehend && the
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: rewrite doesn't need a hyphen
Deprehend is capitalized elsewhere.
write out && as and

File: content/code-generation-from-the-ast.md [hunk 20]
```diff
@@ -0,0 +1,287 @@
...
+
+```bash
+go get -u github.com/deferpanic/deprehend
+go get -u github.com/deferpanic/deferclient
+```
+
+Then run it via:
+
+```bash
+deprehend myproj
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: what is myproj? is it /path/to/myproj? a package name?

File: content/code-generation-from-the-ast.md [hunk 21]
```diff
@@ -0,0 +1,287 @@
...
+
+func main() {
+        go func(){
+                defer deferclient.Persist()
+                bob()
+        }()
+}
+```
+
+This shows that we automatically detect go-routines and re-write the
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: goroutines and rewrite don't need hyphens

File: content/code-generation-from-the-ast.md [hunk 22]
```diff
@@ -0,0 +1,287 @@
...
+source to handle any potential panics that might occur w/in that
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: within

File: content/code-generation-from-the-ast.md [hunk 23]
```diff
@@ -0,0 +1,287 @@
...
+goroutine.
+
+This tool is new and <u>*experimental*</u> and yes, it might eat your cat - so if
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: underlining? hm. 

maybe use `***experimental***` ***experimental*** instead?

File: content/code-generation-from-the-ast.md [hunk 24]
```diff
@@ -0,0 +1,287 @@
...
+you find any bugs or want to extend functionality please create a pull
+request.
+
+We hope this makes developing in go much easier and makes you the end
```
Comment (nathany) [inline content/code-generation-from-the-ast.md]: commas: makes you, the end developer, more productive.

**Comment by nathany** [comment]
Thanks Ian. Interesting post.

File: content/code-generation-from-the-ast.md (hunk 8 above)
Comment (eyberg) [inline content/code-generation-from-the-ast.md]: That's misconstrued - http://golang.org/pkg/debug/elf/#File.Symbols - they are merely stating that it *can* be parsed without a symbol table - some languages you can't do that in w/out one. There are many other things that have to occur besides parsing.

**Comment by eyberg** [comment]
thanks for the review @nathany - fixed - you're one hell of an english teacher ;)

**Comment by nathany** [comment]
Thanks Ian. Looks great.

**Pull request merged**